### PR TITLE
Improvement for imgur links

### PR DESCRIPTION
**Pull request by Vlammar**
Detect if the link is an imgur one, if true try to access directly to the picture, add also a temporary message for case not handeled yet

### CODE REVIEW

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java [hunk 1]
```diff
@@ -68,25 +68,38 @@ static public void render(final URL url, final ImageUtils.ScalingType scaling, f
             @Override
             public ImageMap run() throws Throwable
             {
-                final URLConnection connection = url.openConnection();
+                URL url2=null;
+                if(url.toString().contains("https://imgur.com/")){
```
Comment (AmauryCarrade) [inline src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java]: `startswith` ?

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java [hunk 2]
```diff
@@ -68,25 +68,38 @@ static public void render(final URL url, final ImageUtils.ScalingType scaling, f
...
+                    if(url.toString().contains("gallery/")){
+                        throw new IOException("We do not support imgur gallery yet, please use direct link to image instead (right click on the picture you want to use then select copy link:) ");
```
Comment (AmauryCarrade) [inline src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java]: Extra `:` at the end, and this should be translatable.

```suggestion
                        throw new IOException(I.t("We do not support imgur gallery yet: please use direct link to image instead (right click on the picture you want to use, then select “copy link” or similar)."));
```

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java [hunk 3]
```diff
@@ -68,25 +68,38 @@ static public void render(final URL url, final ImageUtils.ScalingType scaling, f
...
+
+                    }
+                    String newLink="https://i.imgur.com/"+url.toString().split("https://imgur.com/")[1]+".jpg";
```
Comment (AmauryCarrade) [inline src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java]: Should be `final` and could use some sanity checks if the URL does not follow the correct format (else, some `ArrayIndexOutOfRange` exception could be thrown).

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java [hunk 4]
```diff
@@ -68,25 +68,38 @@ static public void render(final URL url, final ImageUtils.ScalingType scaling, f
...
+                    url2=new URL(newLink);
+                }
+
+                if(url2==null)
+                    url2=url;
```
Comment (AmauryCarrade) [inline src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java]: Why? We could use `url` directly, or am I missing something?

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java [hunk 5]
```diff
@@ -102,12 +115,13 @@ public ImageMap run() throws Throwable
...
                 }
 
+
                 if (scaling != ImageUtils.ScalingType.NONE && height <= 1 && width <= 1)
                 {
                     return renderSingle(scaling.resize(image, ImageMap.WIDTH, ImageMap.HEIGHT), playerUUID);
                 }
-
                 final BufferedImage resizedImage = scaling.resize(image, ImageMap.WIDTH * width, ImageMap.HEIGHT * height);
+                //image.flush();
```
Comment (AmauryCarrade) [inline src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java:154]: Forgotten commented code.

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java [hunk 6]
```diff
@@ -126,26 +140,26 @@ public Integer call() throws Exception
         });
 
         final int mapID = futureMapID.get();
-        ImageIOExecutor.saveImage(mapID, image);
+        //ImageIOExecutor.saveImage(mapID, image);
```
Comment (AmauryCarrade) [inline src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java:174]: Forgotten commented code.

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java [hunk 7]
```diff
@@ -126,26 +140,26 @@ public Integer call() throws Exception
...
 
         submitToMainThread(new Callable<Void>()
         {
             @Override
             public Void call() throws Exception
             {
                 Renderer.installRenderer(image, mapID);
+                //image.flush();
```
Comment (AmauryCarrade) [inline src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java:182]: Forgotten commented code.

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java [hunk 8]
```diff
@@ -126,26 +140,26 @@ public Integer call() throws Exception
...
                 return null;
             }
         });
-
+        //image.flush();
```
Comment (AmauryCarrade) [inline src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java:186]: Forgotten commented code.

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java [hunk 9]
```diff
@@ -155,12 +169,10 @@ static private ImageMap renderPoster(final BufferedImage image, final UUID playe
                 return MapManager.getNewMapsIds(mapCount);
             }
         });
-
         poster.splitImages();
-
         final int[] mapsIDs = futureMapsIds.get();
+       // ImageIOExecutor.saveImage(mapsIDs, poster);
```
Comment (AmauryCarrade) [inline src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java:205]: Another forgotten commented code.

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java [hunk 10]
```diff
@@ -178,6 +190,8 @@ public Void call() throws Exception
 
         });
 
+       // image.flush();
```
Comment (AmauryCarrade) [inline src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java:224]: Another forgotten commented code.

File: testforwebhook.test [hunk 11]
```diff
@@ -1 +0,0 @@
-test
```
Comment (AmauryCarrade) [inline testforwebhook.test:1]: What is this for?

**Review by AmauryCarrade — CHANGES_REQUESTED** [review]
Some comments on the code below.

On the implementation: I think we may want a stronger structure than that for this task. For now there's only imgur but we may want to add more automatically converted providers later. What about some real structure for that, with a `ProvidersManager` of some sort and `Provider` classes handling the detection and conversion?

The ultimate objective is to return the “real” URL to test, so such a system could take the user URL and return the real URL, consulting all providers, and returning the original URL if none of the providers match.

Only an idea.

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java (hunk 1 above)
Comment (Vlammar) [reply src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java]: Yeah that's better

**Comment by AmauryCarrade** [comment]
Thanks for your kind consideration.

File: testforwebhook.test (hunk 11 above)
Comment (Vlammar) [reply testforwebhook.test:1]: That was a test when configuring webhooks for discord.

Comment (AmauryCarrade) [reply testforwebhook.test:1]: ~~Should be deleted (not emptied).~~ Nevermind

File: src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java (hunk 4 above)
Comment (Vlammar) [reply src/main/java/fr/moribus/imageonmap/image/ImageRendererExecutor.java]: Outdated i do some code cleanup and removed this.

**Pull request merged**